File: src/streaming/packets.py
"""
Packet dataclasses for the WebSocket server.

Each class represents one JSON message type sent to the dashboard.
Adding a new stream = add a new dataclass here + a loop in ws_server.py.

All packets share a `type` field so the React frontend can route them
with a switch statement.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field


@dataclass
class RawPacket:
    """One second of raw EEG samples, columnar per channel."""
    timestamp: float
    channels:  list[list[float]]   # shape: [n_channels][n_samples]
    type:      str = field(default="raw", init=False)

    def to_json(self) -> str:
        return json.dumps(asdict(self))

@dataclass
class FeaturesPacket:
    timestamp: float
    energy:    float
    focus:     float
    mood:      str
    theta_beta_ratio:          float
    alpha_suppression:         float
    # Attention features from the current mood model. Defaults keep older
    # callers working; warm-up windows emit 0.0 until history fills.
    sustained_attention_index: float = 0.0
    energy_index:              float = 0.0
    is_attentive:              bool  = False
    sustained_streak_sec:      float = 0.0
    type:      str = field(default="features", init=False)

    def to_json(self) -> str:
        return json.dumps(asdict(self))

File: dashboard/src/processing/packets.ts
// Add new packet types here as the server grows.
// Mirror changes in src/streaming/packets.py.

export interface BasePacket {
  type:      string
  timestamp: number
}

export interface RawPacket extends BasePacket {
  type:     'raw'
  channels: number[][]  // columnar: one array of samples per channel
}

export interface FeaturesPacket extends BasePacket {
  type:                        'features'
  energy:                      number
  focus:                       number
  mood:                        string
  theta_beta_ratio:            number
  alpha_suppression:           number
  // Attention features (main_with_signal.py merge); optional for older servers.
  sustained_attention_index?:  number
  energy_index?:               number
  is_attentive?:               boolean
  sustained_streak_sec?:       number
}

export type AnyPacket = RawPacket | FeaturesPacket
